git.Rmd

```diff
@@ -0,0 +1,151 @@
+it. These practices:
+
+* are aspirational but when they can't be followed it's nice to explain why;
+* document heuristics, as detailed instructions may never cover all scenarios;
+* apply to shared commits (maybe after cleaning commits locally).
+
+## tl;dr
+
+* Shared commits^[A _shared_ commit is a commit that someone other than you
+might access, for example, because you pushed it to a remote repository that
+someone else has permission privileges to access.] should not be changed.
+* Changes in each shared commit should be closely related.
+
+## 1. Preserve shared history
+
+Shared commits should never change. Some Git commands preserve the history, but
+others don't. For example, `git commit --amend` and `git rebase --interactive`
+alter the git history; they help to overwrite  a messy commit history with
+a cleaner one; you should use them only before sharing your work. To cleanup
+shared history use instead `git revert`.  If you try to push altered history by
+accident, Git will stop you; to benefit from this protection  avoid `--force`
+(or `-f`) when pushing to shared repositories.
+
+See also:
+
+* [Cleaning _unshared_ history: Squashing all commits in a branch](https://youtu.be/08dhy3Zoob4).
+* [Cleaning _unshared_ history with `git rebase --interactive`](https://youtu.be/cMI8p1XhMzA).
+* [Cleaning _unshared_ history with `git commit --amend`](https://youtu.be/539pfVfr7OI).
+* [Cleaning shared history with `git revert`](https://youtu.be/A8Ld6iDqc3w).
+* [Git playlist](https://www.youtube.com/playlist?list=PLvgdJdJDL-AOHkwiaMvYhPKVjiD9vzZIo).
+
+## 2. Keep commits focused
+
+A commit should focus on related changes only. For example, fixing two
+different bugs should produce two separate commits. Small commits make it
+easier for other developers to understand the changes and roll them back if
+something went wrong. You can avoid sharing unwanted changes via .gitignore. If
+you can't succinctly describe what the commit is doing in a couple of words,
+then it is likely too big.
+
+See also:
+
+* [The right size of a PR](https://github.com/2DegreesInvesting/practices/discussions/3).
+* [.gitignore templates](https://github.com/github/gitignore).
+* Undoing things with Git ([video](https://youtu.be/dZOfEF19yDk),
+[lesson](https://coderefinery.github.io/git-intro/05-undoing)).
+
+## 3. Commit frequently
+
+Committing frequently keeps your commits small. Small commits are more likely
+to be focused; they are easier to review, and easier to merge because they
+offer fewer opportunities for conflicts.
+
+## 4. The work you share should be complete
+
+Each commit and feature should do what it promises, completely; and reverting
+it should undo what that commit or feature meant to do, also completely. If
+the goal of one commit requires changes in two files, that one commit should
+include changes to both files. Similarly, if the goal of a feature requires
+two kinds of changes, that feature should likely include two commits.
```

Comment:
Maybe we can provide an example for the reader of a situation of where a commit does not "do what it promises, completely"? I personally can't think of what that would look like.

Reply:
I think the issue is clearer if we exaggerate your  example. If two commits have the title "Fix all typos" then none does the job completely. That's fine locally but one should thrive to squash those two incomplete commits and instead share only one that really fixes all typos. Do you still think this section is needless?

Reply:
That makes more sense. But it makes more sense in the context of looking back at commits you've already made and saying, "shucks, I should've just put those last few into one commit", versus giving a good rule of thumb to go by when/before making the commits in the first place. I guess you would suggest that if you've noticed something like that, then you should modify your commits/history locally so that they're combined into a single commit before pushing it to the shared repo? I'm not sure that's a great idea... or at least not something we need to encourage people, in particular those who are not super fluent in git (myself included), to do. And ultimately what's the value of that anyway? We push commits to a feature branch, and then ultimately all of those commits get squash-merged into master and the branch is deleted, so.... that seems like more effort than it's worth.

Reply:
The paragraph now includes:
>  For example, if two commits in the same feature branch promise to “Fix all typos”, each
> commit is likely incomplete; they should have been squashed into a single commit before 
> sharing them.

Which rule of thumb would you propose? 

In my opinion, its' okay to encourage people to rewrite their commit history locally -- as much as I encourage rewriting a sentence after drafting it. Few people if any write clear commits or sentences the first try. 

The value of thriving for complete commits in a working feature is that -- even if the feature will be squashed -- each commit can be reverted easily.  It also allows the reviewer to understand the sequence of changes by navigating commits one by one with `n` and `p` under the [commits tab](https://github.com/2DegreesInvesting/practices/pull/7/commits/), rather than the [files changed tab](https://github.com/2DegreesInvesting/practices/pull/7/files).  I find that feature quite usefule in a PR like this one where you may want to see how your comments translated to commits -- for example starting at https://github.com/2DegreesInvesting/practices/pull/7/commits/3897bac0bff0e915919bc9f9afaa23b5a5f6b199 and hitting `n` until the end.

reverting it should undo what that commit or feature meant to do


https://github.com/2DegreesInvesting/practices/pull/7/commits/38b730cb0ab2527e2a79d349f1c5a9d7f733e1bf

Reply:
I can't think of a rule of thumb for this, I guess that's why I brought it up. I'm trying to imagine the perspective when you're making a commit.... you make the changes, stage the changes, then write a commit message when you're committing it that says something like "this commit does XYZ". At that point, chances are that your commit actually does completely do XYZ, unless you accidentally did something unintentional, which almost by definition means you would not be consciously aware enough of it to apply any best practice guidance to avoid it.

So what's the intuition? I think maybe it's not about your commit doing what you said it does, which it probably does unless you made an unintentional mistake. It's about striving to combine like changes into a single commit so that all relevant changes related to a conceptual change are combined into one? So the guidance is something like "before you commit changes, consider if other changes you've made are part of the same conceptual change and combine them into one commit rather than committing them separately"?

Reply:
Thanks! The entire section now reads

> Each commit and feature should do what it promises, completely; and reverting it
should undo what that commit or feature meant to do, also completely. Before you
commit changes, consider if other changes you've made are part of the same
conceptual change and combine them into one commit rather than committing them
separately.

I removed everything else because it seems to be saying more or less the same. But let me know if you thing that some or all of that should stay.

Removed:

> If the goal of one commit requires changes in two files, that one commit should
include changes to both files. For example, if two commits in the same feature
branch promise to "Fix all typos", each commit is likely incomplete; they should
have been squashed into a single commit before sharing them. Similarly, if the
goal of a feature requires two kinds of changes, that feature should likely
include two commits.

Reply:
I like this much shorter version. I'm still not 100% convinced about what it says, but I think it's short and sweet enough to get the general concept across without confounding the topic so much that one is left like 🤔 

I'm also realizing how much my perspective on this and other sections here is impacted by my primary means of git usage (GUI client)... the concept of "staging" multiple and/or subsequent changes is so transparent in my workflow that it practically doesn't exist.